detector/CaloTB/CaloPrototype_v03.cpp

```diff
@@ -0,0 +1,198 @@
+
+    int layer_num = 0;
+    int layerType   = 0;
+
+    double layer_pos_z = - cal_hz;
+
+    for (xml_coll_t c(x_det, _U(layer)); c; ++c) {
+        xml_comp_t x_layer = c;
+        int repeat = x_layer.repeat();                // Get number of times to repeat this layer.
+        const Layer* lay = layering.layer(layer_num); // Get the layer from the layering engine.
+        double layer_thickness = lay->thickness();
+        string layer_type_name   = _toString(layerType,"layerType%d");
+
+        // Loop over repeats for this layer.
+        for (int j = 0; j < repeat; j++) {
+            string layer_name = _toString(layer_num, "layer%d");
+            DetElement layer(layer_name, layer_num);
+
+            // Layer box & volume
+            Volume layer_vol(layer_type_name, Box(cal_hx, cal_hy, layer_thickness / 2), air);
+            
+            // Create the slices (sublayers) within the layer.
+            double slice_pos_z = -(layer_thickness / 2);
+            int slice_number = 0;
+            
+            for (xml_coll_t k(x_layer, _U(slice)); k; ++k) {
+                xml_comp_t x_slice = k;
+                string slice_name = _toString(slice_number, "slice%d");
+                double slice_thickness = x_slice.thickness();
+                Material slice_material = theDetector.material(x_slice.materialStr());
+                DetElement slice(layer, slice_name, slice_number);
+
+                slice_pos_z += slice_thickness / 2;
+                // Slice volume & box
+                Volume slice_vol(slice_name, Box(cal_hx, cal_hy, slice_thickness / 2), slice_material);
+                
+                if (x_slice.isSensitive()) {
+		  sens.setType("calorimeter");
+		  slice_vol.setSensitiveDetector(sens);                   
+                } 
+                
+                // Set region, limitset, and vis.
+                slice_vol.setAttributes(theDetector, x_slice.regionStr(), x_slice.limitsStr(), x_slice.visStr());
+                // slice PlacedVolume
+                PlacedVolume slice_phv = layer_vol.placeVolume(slice_vol, Position(0, 0, slice_pos_z));
+                slice_phv.addPhysVolID("slice", slice_number);
+                slice.setPlacement(slice_phv);
+
+                // Increment Z position for next slice.
+                slice_pos_z += slice_thickness / 2;
+                // Increment slice number.
+                ++slice_number;
+            }
+            
+            // Set region, limitset, and vis.
+            layer_vol.setAttributes(theDetector, x_layer.regionStr(), x_layer.limitsStr(), x_layer.visStr());
+           
+            // Layer position in Z within the stave.
+            layer_pos_z += layer_thickness / 2;
+            // Layer physical volume.
+            PlacedVolume layer_phv = envelope.placeVolume(layer_vol, Position(0, 0, layer_pos_z));
+            layer_phv.addPhysVolID(identifierLayer, layer_num);
```

Comment:
Is this why the driver depends on the segmentation type? Would it be easier to just give that as a string parameter, instead of depending on the segmentation type?

Reply:
Thanks!